core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializer.java

```diff
@@ -0,0 +1,74 @@
+package org.zalando.nakadi.repository.kafka;
+
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DecoderFactory;
+import org.zalando.nakadi.domain.Envelope;
+import org.zalando.nakadi.domain.NakadiRecord;
+import org.zalando.nakadi.service.AvroSchema;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+public class KafkaRecordDeserializer implements RecordDeserializer {
+
+    private final GenericDatumReader nakadiAccessLogReader;
+    private final GenericDatumReader metadataDatumReader;
+    private final ByteArrayOutputStream baos;
+    private final AvroSchema schemas;
+
+    public KafkaRecordDeserializer(final AvroSchema schemas) {
+        this.schemas = schemas;
+        this.metadataDatumReader = new GenericDatumReader(
+                this.schemas.getMetadataSchema());
+        this.nakadiAccessLogReader = new GenericDatumReader(
+                this.schemas.getNakadiAccessLogSchema());
+        this.baos = new ByteArrayOutputStream();
+    }
+
+    public byte[] deserialize(final byte[] eventFormat, final byte[] data) {
+        if (eventFormat == null) {
+            // JSON
+            return data;
+        }
+
+        if (Arrays.equals(eventFormat, NakadiRecord.Format.AVRO.getFormat())) {
+            try {
+                final Envelope envelop = Envelope.deserialize(data);
+                return deserializeAvro(envelop);
```

Comment:
Here we deserialize two times, but if feels like `Envelope` could just hold reference to the underlying byte array, and avoid copying the data.

Reply:
Envelop deserializes message from Kafka bytes to byte array and keeps that array in Envelop. Then the byte array is desirialized to Avro. I am not sure I understand what you mean.

Reply:
I mean in these two lines an instance of `Envelope` holds copy of the `data` bytes array, but strictly speaking, that's not necessary.  Could we avoid making that copy?